modular_citadel/code/modules/reagents/chemistry/recipes/fermi.dm

```diff
@@ -7,98 +7,42 @@
 
 //Called when reaction STOP_PROCESSING
 /datum/chemical_reaction/proc/FermiFinish(datum/reagents/holder, var/atom/my_atom, reactVol)
-	if(clear_conversion == REACTION_CLEAR_IMPURE | REACTION_CLEAR_INVERSE)
+	if(clear_conversion & REACTION_CLEAR_IMPURE | REACTION_CLEAR_INVERSE)
 		for(var/id in results)
-			var/datum/reagent/R = my_atom.reagents.has_reagent("[id]")
+			var/datum/reagent/R = holder.has_reagent("[id]")
+			if(!R) //If we exploded
+				continue
 			if(R.purity == 1)
 				continue
 
-			var/cached_volume = R.volume
-			if(clear_conversion == REACTION_CLEAR_INVERSE && R.inverse_chem)
+			var/temp_purity = R.purity
+			var/cached_volume = reactVol
+			if(clear_conversion & REACTION_CLEAR_INVERSE && R.inverse_chem)
 				if(R.inverse_chem_val > R.purity)
-					my_atom.reagents.remove_reagent(R.id, cached_volume, FALSE)
-					my_atom.reagents.add_reagent(R.inverse_chem, cached_volume, FALSE, other_purity = 1)
-
-			else if (clear_conversion == REACTION_CLEAR_IMPURE && R.impure_chem)
+					if(GLOB.Debug2)
+						message_admins("Inverting [cached_volume] [R.id] [R.purity] into [R.inverse_chem]")
+					holder.remove_reagent(R.id, cached_volume, TRUE)
+					holder.add_reagent(R.inverse_chem, cached_volume, FALSE, added_purity = 1)
+					var/datum/reagent/R2 = holder.has_reagent("[R.inverse_chem]")
+					if(clear_conversion & REACTION_CLEAR_RETAIN)
+						R2.cached_purity = 1-R.purity
+						R2.purity = R2.cached_purity
+					return
+
+			if(clear_conversion & REACTION_CLEAR_IMPURE && R.impure_chem)
 				var/impureVol = cached_volume * (1 - R.purity)
-				my_atom.reagents.remove_reagent(R.id, (impureVol), FALSE)
-				my_atom.reagents.add_reagent(R.impure_chem, impureVol, FALSE, other_purity = 1)
+				if(GLOB.Debug2)
+					message_admins("Impure [cached_volume] of [R.id] at [R.purity] into [impureVol] of [R.impure_chem] with [clear_conversion & REACTION_CLEAR_RETAIN ? "Clear conversion" : "Purification"] mechanics")
```

Comment:
This could use the another macro similar to debug_admins but with message_admins() instead of to_chat()... And maybe a proper reagents logging port at some point smh.

Reply:
I'll atomise this out. I have to rewrite fermichem code to properly done multireactions unless arrested which I'll put up after this is merged.